### DISCUSSION TITLE
How to strongly type a Zustand store?

### DISCUSSION THREAD

**Discussion by EriksonBahr** [pmndrs/zustand#2396] [Q&A]
I'd like to create a strongly type context that returns the instance of the Zustand store ([following this tip](https://github.com/pmndrs/zustand/issues/128#issuecomment-673398578)) but it does not seem that all types are exported. The types `Write` and `StoreSubscribeWithSelector` are nowhere to be found, even though they appear in vscode (see image).

My store looks like this
```typescript
export type AnswersState = {
  state: number
}

export type AnswersMethods = {
  method: () => void
}


export type QuestionnaireState = {
  store: UseBoundStoreWithEqualityFn<Write<StoreApi<AnswersState & AnswersMethods>, StoreSubscribeWithSelector<AnswersState & AnswersMethods>>>
};
```

![image](https://github.com/pmndrs/zustand/assets/4528527/d5a3a585-96c5-4c92-8861-84a483a15e71)

Can anyone give some hints? Thanks!

**Comment by EriksonBahr** [comment]
I ended up copying the unexported types to my project, hopefully there is a better way: 

```typescript
type StoreSubscribeWithSelector<T> = {
  subscribe: {
      (listener: (selectedState: T, previousSelectedState: T) => void): () => void;
      <U>(selector: (state: T) => U, listener: (selectedState: U, previousSelectedState: U) => void, options?: {
          equalityFn?: (a: U, b: U) => boolean;
          fireImmediately?: boolean;
      }): () => void;
  };
};

type Write<T, U> = Omit<T, keyof U> & U;
```

**Reply by dbritto-dev** [reply]
@EriksonBahr hope this helps you out -> https://docs.pmnd.rs/zustand/guides/nextjs
